File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if(lists.length == 0){
            return null;
        }

        ListNode head = null;

        for(int i = 0; i < lists.length; i++){
            head = mergeLists(head, lists[i]);
        }

        return head;
    }

    public ListNode mergeLists(ListNode l1, ListNode l2){
        ListNode head = new ListNode();
        ListNode curr = head;

        while(l1 != null && l2 != null){
            if(l1.val < l2.val){
                curr.val = l1.val;
                l1 = l1.next;
            } else {
                curr.val = l2.val;
                l2 = l2.next;
            }

            curr.next = new ListNode();
            curr = curr.next;
        }

        while(l1 != null){
            curr.val = l1.val;
            l1 = l1.next;
            if(l1 != null){
                curr.next = new ListNode();
                curr = curr.next;
            }
        }

        while(l2 != null){
            curr.val = l2.val;
            l2 = l2.next;
            if(l2 != null){
                curr.next = new ListNode();
                curr = curr.next;
            }
        }

        return head;
    }
}

File: Data Structures & Algorithms/add-two-numbers/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        head = ListNode()
        curr = head

        carry = 0
        while l1 and l2:
            digit = l1.val + l2.val + carry
            carry = digit // 10
            digit %= 10
            curr.val = digit
            l1 = l1.next
            l2 = l2.next
            if l1 or l2 or carry != 0:
                curr.next = ListNode()
                curr = curr.next
        
        while l1:
            digit = l1.val + carry
            carry = digit // 10
            digit %= 10
            curr.val = digit
            l1 = l1.next
            if l1 or carry != 0:
                curr.next = ListNode()
                curr = curr.next

        while l2:
            digit = l2.val + carry
            carry = digit // 10
            digit %= 10
            curr.val = digit
            l2 = l2.next
            if l2 or carry != 0:
                curr.next = ListNode()
                curr = curr.next
        
        if carry != 0:
            curr.val = carry

        return head